tree_hash.go

```diff
@@ -0,0 +1,267 @@
+			return [32]byte{}, err
+		}
+		return merkleize(chunks), nil
+	}
+	return hasher, nil
+}
+
+func makeBasicSliceHasher(typ reflect.Type) (hasher, error) {
+	utils, err := cachedSSZUtilsNoAcquireLock(typ.Elem())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get ssz utils: %v", err)
+	}
+	hasher := func(val reflect.Value) ([32]byte, error) {
+		var serializedValues [][]byte
+		// We encode every serialized value into a list of byte slices.
+		for i := 0; i < val.Len(); i++ {
+			buf := &encbuf{}
+			if err = utils.encoder(val.Index(i), buf); err != nil {
+				return [32]byte{}, err
+			}
+			writer := new(bytes.Buffer)
+			if err = buf.toWriter(writer); err != nil {
+				return [32]byte{}, err
+			}
+			serializedValues = append(serializedValues, writer.Bytes())
+		}
+		chunks, err := pack(serializedValues)
+		if err != nil {
+			return [32]byte{}, err
+		}
+		b := make([]byte, 32)
+		binary.LittleEndian.PutUint64(b, uint64(val.Len()))
+		return mixInLength(merkleize(chunks), b), nil
+	}
+	return hasher, nil
+}
+
+func makeCompositeSliceHasher(typ reflect.Type) (hasher, error) {
+	utils, err := cachedSSZUtilsNoAcquireLock(typ.Elem())
+	if err != nil {
+		return nil, err
+	}
+	hasher := func(val reflect.Value) ([32]byte, error) {
+		roots := [][]byte{}
+		for i := 0; i < val.Len(); i++ {
+			root, err := utils.hasher(val.Index(i))
+			if err != nil {
+				return [32]byte{}, err
+			}
+			roots = append(roots, root[:])
+		}
+		chunks, err := pack(roots)
+		if err != nil {
+			return [32]byte{}, err
+		}
+		b := make([]byte, 32)
+		binary.LittleEndian.PutUint64(b, uint64(val.Len()))
+		return mixInLength(merkleize(chunks), b), nil
+	}
+	return hasher, nil
+}
+
+func makeCompositeArrayHasher(typ reflect.Type) (hasher, error) {
+	utils, err := cachedSSZUtilsNoAcquireLock(typ.Elem())
+	if err != nil {
+		return nil, err
+	}
+	hasher := func(val reflect.Value) ([32]byte, error) {
+		roots := [][]byte{}
+		for i := 0; i < val.Len(); i++ {
+			root, err := utils.hasher(val.Index(i))
+			if err != nil {
+				return [32]byte{}, err
+			}
+			roots = append(roots, root[:])
+		}
+		chunks, err := pack(roots)
+		if err != nil {
+			return [32]byte{}, err
+		}
+		return merkleize(chunks), nil
+	}
+	return hasher, nil
+}
+
+func makeStructHasher(typ reflect.Type) (hasher, error) {
+	fields, err := structFields(typ)
+	if err != nil {
+		return nil, err
+	}
+	hasher := func(val reflect.Value) ([32]byte, error) {
+		roots := [][]byte{}
+		for _, f := range fields {
+			root, err := f.sszUtils.hasher(val.Field(f.index))
+			if err != nil {
+				return [32]byte{}, fmt.Errorf("failed to hash field of struct: %v", err)
+			}
+			roots = append(roots, root[:])
+		}
+		return merkleize(roots), nil
+	}
+	return hasher, nil
+}
+
+func makePtrHasher(typ reflect.Type) (hasher, error) {
+	elemSSZUtils, err := cachedSSZUtilsNoAcquireLock(typ.Elem())
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(1461): The tree-hash of nil pointer isn't defined in the spec.
```

Comment:
What issue does this reference? Something in Prysm? We should transfer it to this repo and use the new issue number if that is the case.